bindings/python/triplestore/backends/ontopy.py

```diff
@@ -0,0 +1,197 @@
+            raise TypeError("`onto` must be either an ontology or None")
+        if load:
+            self.onto.load(**kwargs)
+
+    def triples(self, triple: "Triple") -> "Generator":
+        """Returns a generator over matching triples."""
+
+        def to_literal(o, d):
+            """Returns a literal from (o, d)."""
+            if isinstance(d, str) and d.startswith("@"):
+                lang, datatype = d[1:], None
+            else:
+                lang, datatype = None, d
+            return Literal(o, lang=lang, datatype=datatype)
+
+        s, p, o = triple
+        abb = (
+            None if s is None else self.onto._abbreviate(s),
+            None if p is None else self.onto._abbreviate(p),
+            None if o is None else self.onto._abbreviate(o),
+        )
+        for s, p, o in self.onto._get_obj_triples_spo_spo(*abb):
+            yield (
+                _unabbreviate(self.onto, s),
+                _unabbreviate(self.onto, p),
+                _unabbreviate(self.onto, o),
+            )
+        for s, p, o, d in self.onto._get_data_triples_spod_spod(*abb, d=''):
+            yield (
+                _unabbreviate(self.onto, s),
+                _unabbreviate(self.onto, p),
+                to_literal(o, d),
+            )
+
+    def add_triples(self, triples: "Sequence[Triple]"):
+        """Add a sequence of triples."""
+        for s, p, o in triples:
+            if isinstance(o, Literal):
+                if o.lang:
+                    d = f"@{o.lang}"
+                elif o.datatype:
+                    d = f"^^{o.datatype}"
+                else:
+                    d = 0
+                self.onto._add_data_triple_spod(
+                    self.onto._abbreviate(s),
+                    self.onto._abbreviate(p),
+                    self.onto._abbreviate(o),
+                    d,
+                )
+            else:
+                self.onto._add_obj_triple_spo(
+                    self.onto._abbreviate(s),
+                    self.onto._abbreviate(p),
+                    self.onto._abbreviate(o),
+                )
+
+    def remove(self, triple: "Triple"):
+        """Remove all matching triples from the backend."""
+        s, p, o = triple
+        to_remove = list(self.onto._get_triples_spod_spod(
+            self.onto._abbreviate(s) if s is not None else None,
+            self.onto._abbreviate(p) if s is not None else None,
+            self.onto._abbreviate(o) if s is not None else None,
+        ))
+        for s, p, o, d in to_remove:
+            if d:
+                self.onto._del_data_triple_spod(s, p, o, d)
+            else:
+                self.onto._del_obj_triple_spo(s, p, o)
+
+    # Optional methods
+    def parse(self, source=None, location=None, data=None, format=None,
+              encoding=None, **kwargs):
+        """Parse source and add the resulting triples to triplestore.
+
+        The source is specified using one of `source`, `location` or `data`.
```

Comment:
In the current implementation it will choose the one it finds first in the order source, location, data. This means that if more than one is given it will only use the first it finds. IF that is desired the order of preference should be added in the docstring.

Alternatively, if only one can be given, definig more than one should throw and error.

Reply:
The documentation says "_The source is specified using one of `source`, `location` or `data`._", so I think it is fair to consider providing more than one is undefined behaviour. That leaves little more freedom for the implementation of the backends. OK?

Reply:
OK. But I still think and error or warning should be raised if more than one of them is given.

Reply:
A check has been pushed